### PR TITLE
[FIX] clear move_name when unreconciling bank statement line

### PR DESCRIPTION
**Pull request by yostashiro**
### Description of the issue/feature this PR addresses:

It is, by design, not possible to delete bank statement line when a value remains in move_name field of bank statement line (see `unlink` method in this file).

### Current behavior before PR:

Current code does not clear the move_name field even when related journal entries are cancelled, which causes the problem of not being able to delete the unreconciled bank statement line.

### Desired behavior after PR is merged:

Clears move_name field when bank statement line is unreconciled, so that it is possible to delete the line.

@qrtl

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tivisse** [comment]
@yostashiro The constraint on the unlink method has been removed since https://github.com/odoo/odoo/commit/126df41f3d99cdfda6918092d5fa231936ac17f6

This shouldn't be a blocking point for you now. So I cancel this PR.

Thanks for the feedback.

**Comment by yostashiro** [comment]
@tivisse Thanks for the update, however I do not think the current design with https://github.com/odoo/odoo/commit/126df41f3d99cdfda6918092d5fa231936ac17f6 is sufficient.

For example, it creates confusing situation in following scenario (which is quite common, I believe):

1. Import a bank statement.  It contains a line for bank charge.
2. Reconcile this line through Bank Reconciliation screen by creating a new expense journal entry.
3. User realizes that there was some mistake in Step 2 (e.g. forgot to key in analytic account), so he/she opens the journal entry from Step 2, cancel and delete it, which unreconciles the statement line.
4. User tries to reconcile the statement line again, but it is not possible because it has a move_name which no longer exists assigned.

Maybe you don't expect that user does such operation as Step 3 (I guess you expect that user should just unreconcile the statement line to remove the generated journal entry, instead of deleting the journal entry before unreconciling the statement line), but user would naturally make this mistake without knowing the details of Odoo's design.

Therefore, what I think would be good is that Odoo blocks cancellation of journal entries which are generated from bank reconciliation if `move_name` of the statement line is not cleared with deletion of the journal entry (I couldn't understand this comment "However, the reference is useful to reuse the name if the line is reconciled again" in this commit https://github.com/odoo/odoo/commit/126df41f3d99cdfda6918092d5fa231936ac17f6, by the way).

**Comment by tivisse** [comment]
Hi @yostashiro 

We try to assign the community PR to the related teams now. I'm gonna let someone of the Accounting team manage this for now. 

Thanks for the answer and have a nice day.

**Comment by yostashiro** [comment]
@tivisse Thanks for reopening this PR.  My Enterprise customer has been troubled by the current behavior and there have been many cases of them not being able to re-reconcile the statement line because of the uncleared move_name.  I have been forced to clear the move_name field manually through Postgres every time this happens, so the situation is far from ideal.  Would be good to have a clean solution.

**Comment by nim-odoo** [comment]
That seems to go against [the purpose](https://github.com/odoo/odoo/blob/10.0/addons/account/models/account_bank_statement.py#L375) of this field:

> Technical field holding the number given to the journal entry, automatically set when the statement line is reconciled then stored to set the same number again if the line is cancelled, set to draft and re-processed again.

That would probably make sense to clean it if the move is unlinked, though.

**Comment by nim-odoo** [comment]
@mart-e why did you revert https://github.com/odoo/odoo/commit/6de2990f8d45074d39917ab132791b3ddc7e6e30?
That still makes sense to me to clean-up the move_name if the move is deleted.

**Comment by mart-e** [comment]
@nim-odoo I reverted after discussion with @qdp-odoo
 I explain at 126df41f3d99cdfda6918092d5fa231936ac17f6.
TL;DR : it is on purpose to reuse the move_name to avoid gaps.

**Comment by nim-odoo** [comment]
After discussion with @qdp-odoo , none of the solutions suggested here (including mine) should be merged.

The issue arises if, at step 4. of https://github.com/odoo/odoo/pull/15715#issuecomment-318828452 , you try to reconcile with an existing move line (let's call it A). If you create a new move line, it works, and the line created will receive the move name of step 2.

The workaround would be to reverse A. That would prevent any gap, but I agree that it's not very user friendly. It you don't mind the sequence number gaps, I would apply a local patch, such as your or [mine](https://github.com/odoo-dev/odoo/commit/0b34d78b83f3f359ad270523720a27bdf1cd2c95). Since Odoo is built to prevent gaps, there is nothing that can be done in a standard branch.

**Comment by cliffkujala** [comment]
So what is the solution as a user to this then?
If I find a mistake where we created new entry from bank statement instead of selecting the correct internal transfer, I don't know how to fix it.  If I un-reconcile the line so that I can reselect the correct internal transfer it breaks the entire statement, and I have to reconcile the entire statement all over again.  This can't be right. Users will make mistakes so there must be a way to back them out.

**Comment by ray-odoo** [comment]
@cliffkujala Currently allowing users to remove the value in move_name by creating a Menu for Bank Statement Lines. The tree view needs editable="top" and the move_name field added with readonly="0" and then users can easily find and edit entries that need to be reconciled after mistakes.
